### DISCUSSION TITLE
Support for SignalR

### DISCUSSION THREAD

**Discussion by gentledepp** [ChilliCream/hotchocolate#4851] [Feature Request]
**Is your feature request related to a problem? Please describe.**
I am wondering how connection management works with hotchocolate (websockets - what if a connection is lost, etc)
As GraphQL is transport agnostic, I believe using SignalR as a transport would be a great option as
  - it is a well established .net core communication framework and handles connection interruptions out-of-the-box
  - it can easily be scaled using Azure SignalR

**Describe the solution you'd like**
You telling me that your websockets implementation can do all of the above or that SignalR is already a supported transport option.

**Additional context**
I am learning about hotchocolate and straberry??? and was especially intrigued by the async enumerable support for server-side events. So I thought I may implement some chat application to test this out. However, when I heard "web sockets" I was immediately thinking about how to scale out this chat app in theory. This, in turn, led me to the question why SignalR was not a supported transport channel.

Please enlighten me! :-)

**Comment by michaelstaib** [comment]
SignalR is on our list, but we are first implementing the new `graphql-ws` protocol, which is being specified with the GraphQL over HTTP spec. The protocol in general handles connection aborts since you can reconnect using the subscription id. The server can also kill connections so that clients reconnect to another server and get their events without disruption.

Before we do SignalR, we also want to add gRPC, which also has nice streaming capabilities and helps us more when doing schema stitching.

Regarding SignalR, we think it will be inconcluded with the second feature update to Hot Chocolate 11.

**Comment by michaelstaib** [comment]
https://github.com/graphql/graphql-over-http/pull/140

**Comment by gentledepp** [comment]
Ok, thanks for the info!
I feel enlightened ;-)

**Comment by p-bojkowski** [comment]
What do you think about the new [Azure Web PubSub](https://azure.microsoft.com/en-us/services/web-pubsub/) ?
For Subscriptions?  :)

**Comment by douglasg14b** [comment]
The azure web pub/sub is... Excessively expensive is what it is. And it hope it doesn't become the target over more native support. 

I see v12 has some websockets stuff, does this mean it's one step closer to SignalR support?

Edit: If you had 1000 concurrent users online 24/7 on average, JUST the SignalR "keep-alive" every 10 seconds would cost you $259.2/month. Which is _**beyond**_ insane. `1000 * 6 * 60 * 24 * 30 / 1000000`. $0.01/10,000 messages is an astronomical price for a chatty, realtime, application.

**Reply by michaelstaib** [reply]
You can use Hot Chocolate without Azure Web PubSub just with the built-in WebSocket support. We support websockets now for years. At the moment there are no clients in the GraphQL ecosystem using SignalR. For version 13 we are implementing the new graphql-ws protocol and the graphql-sse support. 

https://github.com/enisdenjo/graphql-ws
https://github.com/enisdenjo/graphql-sse

**Reply by syedqutub** [reply]
how can i setup the HotChocolate server to use SSE as a transport for subscription? or is this still in development?

**Comment by douglasg14b** [comment]
@PascalSenn  Why converted to a discussion?

**Reply by PascalSenn** [reply]
We moved a bunch of feature requests to discussions to cleanup the isssue board. We have a lot of old and already resolved issues there, and for feature requests the discussions seems to fit better.

**Comment by sturlath** [comment]
I´m trying to find information on using Hotchocolate on top of my grpc endpoints but ain't finding much on it except this https://github.com/ChilliCream/hotchocolate/pull/1122

Some blog post or example would be nice if anybody knows about some!

**Comment by jfheins** [comment]
We use SignalR primarily for the feature that it can switch protocol. We are behind Azure FrontDoor, so for all our testing environments, Websockets are unavailable. But we rely on SignalR to switch so that we test our software with Long-Polling and then in the later stages of testing it will use WebSockets seamlessly. 
Using GraphQL over SignalR would be ideal :-)

**Reply by michaelstaib** [reply]
what about graphql-sse.

The thing why we have not moved yet in the direction of SignalR is that there is no usage of it outside the .NET ecosystem. The GraphQL community at the moment embrace two standards ... graphql-ws (web sockets) and graphql-sse (Server Side Events) the latter should also work with Azure FrontDoor.

**Reply by michaelstaib** [reply]
https://github.com/enisdenjo/graphql-sse

**Reply by jfheins** [reply]
I think SSE might work 👍 At least when I give SignalR the option to use it, it works.
Looking forward to try out HC subscriptions via SSE then :-)
